File: NetDeck/Hypergeometric.m
//
//  Hypergeometric.m
//  Net Deck
//
//  Created by Gereon Steffens on 06.03.14.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "Hypergeometric.h"

@interface Binominal: NSObject
@end

@implementation  Binominal

// calculate the binominal coefficient (n over k)

+(unsigned long long) coefficientWithN:(NSUInteger)n over:(NSUInteger)k
{
    if (k == 0)
    {
        return 1;
    }
    if (n == 0 || k > n)
    {
        return 0;
    }

    if ((2*k) > n)
    {
        return [Binominal coefficientWithN:n over:n-k];
    }
    else
    {
        unsigned long long result = n - k + 1;
        for (int i=2; i<=k; ++i)
        {
            result *= (n-k+i);
            result /= i;
        }
        return result;
    }
}

@end

@implementation Hypergeometric

+(double) getProbabilityFor:(NSUInteger)desiredCards cardsInDeck:(NSUInteger)cardsInDeck desiredCardsInDeck:(NSUInteger)desiredCardsInDeck cardsDrawn:(NSUInteger)cardsDrawn
{
    double r = 0;
    
    for (; desiredCards<=cardsDrawn; desiredCards++)
    {
        r += [Hypergeometric get:desiredCards cardsInDeck:cardsInDeck desiredCardsInDeck:desiredCardsInDeck cardsDrawn:cardsDrawn];
    }
    return MIN(r, 1.0);
}

+(double) get:(NSUInteger)desiredCards cardsInDeck:(NSUInteger)cardsInDeck desiredCardsInDeck:(NSUInteger)desiredCardsInDeck cardsDrawn:(NSUInteger)cardsDrawn
{
    if (desiredCards==0 || cardsInDeck==0 || desiredCardsInDeck==0 || cardsDrawn==0)
    {
        return 0;
    }

    double d = [Binominal coefficientWithN:cardsInDeck over:cardsDrawn];
    if (d == 0)
    {
        return 0;
    }

    double b1 = [Binominal coefficientWithN:desiredCardsInDeck over:desiredCards];
    double b2 = [Binominal coefficientWithN:cardsInDeck-desiredCardsInDeck over:cardsDrawn-desiredCards];
    return b1 * b2 / d;
}


@end

File: NetDeck/Hypergeometric.swift
//
//  Hypergeometric.swift
//  NetDeck
//
//  Created by Gereon Steffens on 13.12.15.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import Foundation

final class Hypergeometric {
    static func getProbabilityFor(_ desiredCards: Int, cardsInDeck: Int, desiredCardsInDeck: Int, cardsDrawn: Int) -> Double {
        assert(desiredCards>0 && cardsInDeck>0 && desiredCardsInDeck>0 && cardsDrawn>0)
        var r = 0.0
        for dc in desiredCards ... cardsDrawn {
            r += probabilityFor(dc, cardsInDeck, desiredCardsInDeck, cardsDrawn)
        }
        return min(r, 1.0)
    }
    
    private static func probabilityFor(_ desiredCards: Int, _ cardsInDeck: Int, _ desiredCardsInDeck: Int, _ cardsDrawn: Int) -> Double {
        if desiredCards==0 || cardsInDeck==0 || desiredCardsInDeck==0 || cardsDrawn==0 {
            return 0.0
        }
        let d = coefficient(for: cardsInDeck, over: cardsDrawn)
        if d == 0.0 {
            return 0.0
        }
        
        let b1 = coefficient(for: desiredCardsInDeck, over: desiredCards)
        let b2 = coefficient(for: cardsInDeck-desiredCardsInDeck, over: cardsDrawn-desiredCards)
        return b1 * b2 / d
    }
    
    private static func coefficient(for n: Int, over k: Int) -> Double {
        assert(n>=0 && k>=0)
        
        if k == 0 { return 1.0 }
        if n == 0 || k > n { return 0.0 }
        
        if 2*k > n {
            return coefficient(for: n, over: n-k)
        } else {
            var result = Double(n - k + 1)
            if k > 1 {
                for i in 2 ... k {
                    result *= Double(n - k + i)
                    result /= Double(i)
                }
            }
            return result
        }
    }
}
